lib/Conversion/TosaToXTenNN.cpp

```diff
@@ -0,0 +1,225 @@
+                                         "expected to be a constant.");
+    }
+
+    // We want to make sure we have QDQ operations surrounded by MULs.
+    Operation *mulOperandOp = mulOp->getOperand(0).getDefiningOp();
+    APFloat quantizeScaleFactor(0.0);
+    auto isQDQPattern = m_Op<amd::xten_nn::DequantizeOp>(
+        m_Op<amd::xten_nn::QuantizeOp>(m_Op<tosa::MulOp>(
+            matchers::m_Any(), m_ConstantFloat(&quantizeScaleFactor))));
+    if (mulOperandOp == nullptr || !isQDQPattern.match(mulOperandOp)) {
+      return rewriter.notifyMatchFailure(mulOp->getLoc(),
+                                         "expected mul->q->dq->mul pattern.");
+    }
+    auto dequantizeOp = cast<amd::xten_nn::DequantizeOp>(mulOperandOp);
+
+    // The quantize operation will have it's reciprocal value constant folded.
+    // So we need to calculate one over to get the scale factor back.
+    APFloat recipValue((float)1.0);
+    recipValue.divide(quantizeScaleFactor, APFloat::rmNearestTiesToEven);
+    if (recipValue != dequantizeScaleFactor) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "expected constants of both multiplications around "
+                          "the QDQ to be equal.");
+    }
+
+    // Only power of two values are supported by the QDQ operations
+    std::optional<int32_t> scaleFactor =
+        getLog2Value(dequantizeScaleFactor.convertToFloat());
+    if (!scaleFactor.has_value()) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "constant is not a integer log2 value.");
+    }
+
+    auto quantizeOp = cast<amd::xten_nn::QuantizeOp>(
+        dequantizeOp->getOperand(0).getDefiningOp());
+
+    // Sum the shifts of the quantize, dequantize and update the operations
+    llvm::APInt shiftSum(32, dequantizeOp.getShift(), true);
+    bool overflow = false;
+    llvm::APInt scaleFactorInt(32, scaleFactor.value(), true);
+    shiftSum = shiftSum.sadd_ov(scaleFactorInt, overflow);
+    if (overflow) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "Adding the shifts of the mul and QDQ overflowed.");
+    }
+    quantizeOp.setShift((int32_t)shiftSum.getSExtValue());
+    dequantizeOp.setShift((int32_t)shiftSum.getSExtValue());
+
+    // Remove the multiplications around the QDQ
+    auto *quantizeMulOp = quantizeOp->getOperand(0).getDefiningOp();
+    quantizeOp->setOperand(0, quantizeMulOp->getOperand(0));
+    rewriter.replaceOp(mulOp, dequantizeOp->getResults());
+    rewriter.eraseOp(quantizeMulOp);
+
+    return success();
+  }
+};
+
+class TosaToXTenNNPass
+    : public xilinx::xten::TOSAToXTenNNBase<TosaToXTenNNPass> {
+public:
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+    MLIRContext *context = module.getContext();
+    RewritePatternSet patterns(context);
+
+    // Ensures constants on the add, mul, sub are on the RHS
+    populateCommutativityUtilsPatterns(patterns);
```

Comment:
Did this work for you? When @TinaAMD and I looked at it, the pass didn't do anything.